src/main/java/com/uber/cadence/internal/sync/ActivityExecutionContextImpl.java

```diff
@@ -45,28 +49,88 @@
   private static final Logger log = LoggerFactory.getLogger(ActivityExecutionContextImpl.class);
 
   private final IWorkflowService service;
-
   private final String domain;
-
   private final ActivityTask task;
   private final DataConverter dataConverter;
   private boolean doNotCompleteOnReturn;
+  private final long heartbeatIntervalNanos;
+  private Object lastDetails;
+  private final ScheduledExecutorService heartbeatExecutor;
+  private Lock lock = new ReentrantLock();
+  private ScheduledFuture future;
+  private ActivityCompletionException lastException;
 
   /** Create an ActivityExecutionContextImpl with the given attributes. */
   ActivityExecutionContextImpl(
-      IWorkflowService service, String domain, ActivityTask task, DataConverter dataConverter) {
+      IWorkflowService service,
+      String domain,
+      ActivityTask task,
+      DataConverter dataConverter,
+      ScheduledExecutorService heartbeatExecutor) {
     this.domain = domain;
     this.service = service;
     this.task = task;
     this.dataConverter = dataConverter;
+    this.heartbeatIntervalNanos = (long) (0.8 * task.getHeartbeatTimeout().toNanos());
+    this.heartbeatExecutor = heartbeatExecutor;
   }
 
-  /**
-   * @throws CancellationException
-   * @see ActivityExecutionContext#recordActivityHeartbeat(Object)
-   */
+  /** @see ActivityExecutionContext#recordActivityHeartbeat(Object) */
   @Override
   public void recordActivityHeartbeat(Object details) throws ActivityCompletionException {
+    lock.lock();
+    try {
+      if (future == null) {
+        doHeartBeat(details);
+      } else {
+        lastDetails = details;
+      }
+      if (lastException != null) {
+        throw lastException;
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void doHeartBeat(Object details) {
+    try {
+      sendHeartbeatDetails(details);
+      onHeartbeatSucceed();
+    } catch (TException e) {
+      // Not rethrowing to not fail activity implementation on intermittent connection or Cadence
```

Comment:
Shouldn't it keep retrying in case of intermittent errors even if recordActivityHeartbeat is not called again?